agent/consul/discoverychain/compile.go

```diff
@@ -304,13 +301,58 @@ func (c *compiler) compile() (*structs.CompiledDiscoveryChain, error) {
 	}, nil
 }
 
-func (c *compiler) detectCircularSplits() error {
-	// TODO(rb): detect when a tree of splitters backtracks
-	return nil
-}
+func (c *compiler) detectCircularReferences() error {
+	var (
+		todo       stringStack
+		visited    = make(map[string]struct{})
+		visitChain stringStack
+	)
+
+	todo.Push(c.startNode)
+	for {
+		current, ok := todo.Pop()
+		if !ok {
+			break
+		}
+		if current == "_popvisit" {
+			if v, ok := visitChain.Pop(); ok {
+				delete(visited, v)
+			}
+			continue
+		}
+		visitChain.Push(current)
+
+		if _, ok := visited[current]; ok {
+			return &structs.ConfigEntryGraphError{
+				Message: fmt.Sprintf(
+					"detected circular reference: [%s]",
+					strings.Join(visitChain.Items(), " -> "),
+				),
+			}
+		}
+		visited[current] = struct{}{}
+
+		todo.Push("_popvisit")
+
+		node := c.nodes[current]
+
+		switch node.Type {
+		case structs.DiscoveryGraphNodeTypeRouter:
+			for _, route := range node.Routes {
+				todo.Push(route.NextNode)
+			}
+		case structs.DiscoveryGraphNodeTypeSplitter:
+			for _, split := range node.Splits {
+				todo.Push(split.NextNode)
+			}
+		case structs.DiscoveryGraphNodeTypeResolver:
+			// Circular redirects are detected elsewhere and failover isn't
```

Comment:
🤔 isn't it convenient to be able to detect cycles in the whole graph from this one function? Even if we do that else where might there be cases in future where we'd want to just be able to validate a graph with one pass? Maybe I'm missing something about why it's more correct or better to split redirects out?

Reply:
Redirects are compiled out during normal graph construction. Redirects as a concept never make it into the final product, so they're the one part that has to be detected during construction rather than after.